durable-storage/src/persistence_layer.rs

```diff
@@ -149,23 +182,68 @@ impl PersistenceLayer {
         })
     }
 
+    /// Whenever a mutable operation is performed, we need to ensure that the persistence layer is
+    /// in [`Mode::Temporary`] mode.
+    fn make_temporary(&mut self) -> Result<(), Error> {
+        match &self.mode {
+            Mode::Temporary { .. } => (),
+            Mode::FromCommit { repo } => {
+                *self = self.try_clone(repo)?;
+            }
+        };
```

Comment:
I would have thought Clippy yells at you for not making this an `if-let` statement. 

Nit but a tiny bit more readable:

```suggestion
        if let Mode::FromCommit { repo } = &self.mode {
            *self = self.try_clone(repo)?;
        };
```